examples/zero_copy/c++/zero_copy_subscriber.cxx

```diff
@@ -0,0 +1,497 @@
+
+struct CommandLineArguments {
+public:
+    int domain_id;
+    int size;
+    int sample_count;
+    bool verbose;
+
+    CommandLineArguments() {
+        domain_id  = 0;
+        sample_count = 1200;
+        size = 1048576;
+        verbose = false;
+    }
+};
+CommandLineArguments arg;
+
+// Get the shared memory key from the publication data property
+int lookup_shmem_key_property(DDS::PropertyQosPolicy& property)
+{
+    const DDS::Property_t *p =
+            DDS::PropertyQosPolicyHelper::lookup_property(property, "shmem_key");
+    if (p == NULL) {
+        return -1;
+    }
+
+    int key;
+    std::istringstream str_to_key(p->value);
+    str_to_key >> key;
+
+    return key;
+}
+
+bool operator<(const DDS::GUID_t& a, const DDS::GUID_t& b)
+{
+    return DDS_GUID_compare(&a, &b) < 0;
+}
+
+bool operator==(const DDS::GUID_t& a, const DDS::GUID_t& b)
+{
+    return DDS_GUID_equals(&a, &b) == DDS_BOOLEAN_TRUE;
+}
+
+class ShmemKeyTrackingReaderListener : public DDSDataReaderListener {
+
+public:
+    ~ShmemKeyTrackingReaderListener()
+    {
+        for (FrameSetViewMap::iterator iter = frame_set_views.begin();
+                iter != frame_set_views.end(); ++iter) {
+            delete iter->second;
+        }
+    }
+
+    FrameSetView* get_frame_set_view(const DDS::GUID_t& publication_guid) const
+    {
+        FrameSetViewMap::const_iterator key_it = frame_set_views.find(
+                publication_guid);
+        if (key_it == frame_set_views.end()) {
+            return NULL;
+        }
+
+        return key_it->second;
+    }
+
+    size_t get_frame_set_view_count() const
+    {
+        return frame_set_views.size();
+    }
+
+private:
+    typedef std::map<DDS::GUID_t, FrameSetView*> FrameSetViewMap;
+
+    // Map to store the frame_set_views created for each matched frame_set
+    FrameSetViewMap frame_set_views;
+
+    void on_subscription_matched(
+        DDSDataReader* reader,
+        const DDS_SubscriptionMatchedStatus& status) /* override */
+    {
+        try {
+            DDS::PublicationBuiltinTopicData publication_data;
+            if (reader->get_matched_publication_data(
+                    publication_data,
+                    status.last_publication_handle) != DDS_RETCODE_OK) {
+                return;
+            }
+
+            int key = lookup_shmem_key_property(publication_data.property);
+
+            if (key == -1) {
+                return;
+            }
+
+            // Create the frame_set_view with the key and size. This frame set
+            // view attaches to the shared memory segment created by the frame
+            // set managed by the matched DataWriter
+            FrameSetView *frame_set_view = new FrameSetView(key, arg.size);
+            DDS::GUID_t publication_guid;
+            DDS_BuiltinTopicKey_to_guid(
+                    &publication_data.key,
+                    &publication_guid);
+
+            frame_set_views[publication_guid] = frame_set_view;
```

Comment:
If the guid is duplicate, this will overwrite the existing pointer and leak it. Should check first if the key exists (frame_set_views.find(publication_guid) == frame_set_views.end()).

I guess that will only happen if the event signals that the pub is being disposed (event count change = -1)